### PR TITLE
[GLUTEN-9540][FLINK] Add UT for filter & project operators

### PR DESCRIPTION
**Pull request by yuanhang-dev**
## What changes were proposed in this pull request?

As discussed in https://github.com/apache/incubator-gluten/pull/9630#pullrequestreview-2965210597, this PR will add UT for filter & project operators

(Fixes: \#ISSUE-ID)

## How was this patch tested?

CI

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by yuanhang-dev** [comment]
@PHILO-HE

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+              new IntType(), new VarCharType(VarCharType.MAX_LENGTH), new IntType()
+            },
+            new String[] {"id", "name", "age"});
+    List<String> inNames = Utils.getNamesFromRowType(flinkRowType);
+
+    RexNode ageFieldRef =
+        rexBuilder.makeInputRef(typeFactory.createSqlType(SqlTypeName.INTEGER), 2);
+    RexNode literal18 =
+        rexBuilder.makeLiteral(
+            new BigDecimal(18), typeFactory.createSqlType(SqlTypeName.INTEGER), false);
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: why use BigDecimal?

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        RowType.of(
+            new LogicalType[] {
+              new IntType(), new VarCharType(VarCharType.MAX_LENGTH), new IntType()
+            },
+            new String[] {"id", "name", "age"});
+    List<String> inNames = Utils.getNamesFromRowType(flinkRowType);
+
+    RexNode ageFieldRef =
+        rexBuilder.makeInputRef(typeFactory.createSqlType(SqlTypeName.INTEGER), 2);
+    RexNode literal18 =
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: Name suggestion: age

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  }
+
+  @Override
+  @Test
+  public void testOpenClose() throws Exception {
+    RowType flinkRowType = RowType.of(new LogicalType[] {new IntType()}, new String[] {"id"});
+    List<String> inNames = Utils.getNamesFromRowType(flinkRowType);
+
+    RexNode idFieldRef = rexBuilder.makeInputRef(typeFactory.createSqlType(SqlTypeName.INTEGER), 0);
+    RexNode literal0 =
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: name suggestion: id

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+            },
+            new String[] {"id", "name", "age"});
+    List<String> inNames = Utils.getNamesFromRowType(flinkRowType);
+
+    RexNode ageFieldRef =
+        rexBuilder.makeInputRef(typeFactory.createSqlType(SqlTypeName.INTEGER), 2);
+    RexNode literal18 =
+        rexBuilder.makeLiteral(
+            new BigDecimal(18), typeFactory.createSqlType(SqlTypeName.INTEGER), false);
+    RexNode filterConditionFlink =
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: flinkFilterCondition

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  @Test
+  public void testOpenClose() throws Exception {
+    RowType flinkRowType = RowType.of(new LogicalType[] {new IntType()}, new String[] {"id"});
+    List<String> inNames = Utils.getNamesFromRowType(flinkRowType);
+
+    RexNode idFieldRef = rexBuilder.makeInputRef(typeFactory.createSqlType(SqlTypeName.INTEGER), 0);
+    RexNode literal0 =
+        rexBuilder.makeLiteral(
+            new BigDecimal(0), typeFactory.createSqlType(SqlTypeName.INTEGER), false);
+    RexNode filterConditionFlink =
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: flinkFilterCondition

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        actualOutput.add(((StreamRecord<RowData>) record).getValue());
+      }
+    }
+
+    List<RowData> expectedOutput =
+        Arrays.asList(
+            GenericRowData.of(1, StringData.fromString("Alice"), 20),
+            GenericRowData.of(3, StringData.fromString("Charlie"), 25));
+
+    assertThat(actualOutput).hasSize(expectedOutput.size());
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: Maybe, also add tests for LessThan, EqualTo. We can then move the result check to a helper method for reuse.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    @Override
+    public void open() throws Exception {
+      if (closed) {
+        throw new IllegalStateException("Close called before open.");
+      }
+      super.open();
+      opened = true;
+      LOG.debug("GlutenSingleInputOperator opened successfully");
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: Remove this log since the flag will be checked.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    @Override
+    public void close() throws Exception {
+      if (!opened) {
+        throw new IllegalStateException("Open was not called before close.");
+      }
+      super.close();
+      closed = true;
+      LOG.debug("GlutenSingleInputOperator closed successfully");
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: Ditto.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+public class GlutenStreamFilterTest extends StreamFilterTest {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GlutenStreamFilterTest.class);
+  private static FlinkTypeFactory typeFactory;
+  private static RexBuilder rexBuilder;
+
+  @BeforeAll
+  public static void setupAll() {
+    LOG.info("GlutenStreamFilterTest setup");
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: This log looks verbose, can be removed.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamProjectTest.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class GlutenStreamProjectTest extends StreamProjectTest {
+  private static final Logger LOG = LoggerFactory.getLogger(GlutenStreamProjectTest.class);
+  private static FlinkTypeFactory typeFactory;
+  private static RexBuilder rexBuilder;
+
+  @BeforeAll
+  public static void setupAll() {
+    LOG.info("GlutenStreamProjectTest setup");
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamProjectTest.java]: Ditto.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Queue;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class GlutenStreamFilterTest extends StreamFilterTest {
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: Is it necessary to extend `StreamFilterTest`? Seems all test methods in `StreamFilterTest` have been overridden.

**Review by PHILO-HE — COMMENTED** [review]
Some comments. Thanks!

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java (hunk 11 above)
Comment (yuanhang-dev) [reply gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: It is indeed unnecessary. The reason for extending `StreamFilterTest` here is that in previous discussions, it was hoped that the unit tests for gluten-flink would be based on vanilla Flink. Do you think we need to remove `extend`?

**Comment by shuai-xu** [comment]
Thank you for the work. After #https://github.com/apache/incubator-gluten/pull/9940 , the gluten operator changed to GlutenVector***Operator to support passing RowVector with changing to RowData, so please test the Vector based operator  first.

File: gluten-flink/ut/pom.xml [hunk 12]
```diff
@@ -44,7 +43,7 @@
             <artifactId>gluten-flink-runtime</artifactId>
             <version>${project.version}</version>
         </dependency>
-
```
Comment (philo-he) [inline gluten-flink/ut/pom.xml:47]: Nit: no need to keep one empty line between dependencies. Please remove it.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+
+    assertThat(harness.getOutput()).isNotEmpty();
+
+    harness.close();
+
+    assertThat(operator.isOpened()).isTrue();
+    assertThat(operator.isClosed()).isTrue();
+  }
+
+  private void testComparisonFilter(
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: testFilter

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+        rexBuilder.makeInputRef(typeFactory.createSqlType(SqlTypeName.VARCHAR), 1);
+    RexNode name =
+        rexBuilder.makeLiteral("Alice", typeFactory.createSqlType(SqlTypeName.VARCHAR), false);
+    RexNode flinkFilterCondition =
+        rexBuilder.makeCall(SqlStdOperatorTable.EQUALS, nameFieldRef, name);
+
+    List<RowData> expectedOutput =
+        Arrays.asList(GenericRowData.of(1, StringData.fromString("Alice"), 20));
+
+    testFilterWithCondition(flinkFilterCondition, expectedOutput);
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: I note some function calls this to filter name and some calls testComparisonFilter to filter age. It may be better to use a unified util function called `testFilter` with `RexNode flinkFilterCondition` as its argument. The `flinkFilterCondition` is constructed at caller side.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+  }
+
+  private void assertRowDataListEquals(List<RowData> actual, List<RowData> expected) {
+    assertThat(actual).hasSize(expected.size());
+    for (int i = 0; i < expected.size(); i++) {
+      assertRowDataEquals(actual.get(i), expected.get(i));
+    }
+  }
+
+  private void assertRowDataEquals(RowData actual, RowData expected) {
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/GlutenStreamFilterTest.java]: naming suggestion: `checkEquals`

Please check Flink code to see whether it has a public compare method that we can directly use.
 
It would be better to make this function general (maybe, move it to a util class for other tests to use), not bound to this test class. For general case, we don't know each column's data type beforehand. Given any RowData, we can get each column's data type and using the corresponding getter method to get the data for comparison.

For strict comparison, please check whether they have same fields number obtained by `getArity()`.
Please also make sure null data is allowed in this check. It would be better to add some test cases with NULL data.

**Review by PHILO-HE — COMMENTED** [review]
Some comments. Thanks!

**Comment by yuanhang-dev** [comment]
@shuai-xu @PHILO-HE Could you please review the code again? Thank you

**Comment by shuai-xu** [comment]
> @shuai-xu @PHILO-HE Could you please review the code again? Thank you

Will you add tests for GlutenVectorXXXOperator in this pr or a seperated pr?

**Comment by yuanhang-dev** [comment]
> > @shuai-xu @PHILO-HE Could you please review the code again? Thank you
> 
> Will you add tests for GlutenVectorXXXOperator in this pr or a seperated pr?

I will add tests in another PR after the q4 PR is merged.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    for (int i = 0; i < actual.getArity(); i++) {
+      RowData.FieldGetter getter = RowData.createFieldGetter(fieldTypes.get(i), i);
+      Object actualValue = getter.getFieldOrNull(actual);
+      Object expectedValue = getter.getFieldOrNull(expected);
+
+      assertEquals("Field " + i + " mismatch", expectedValue, actualValue);
+    }
+  }
+
+  public static void assertRowDataListEquals(
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java]: Also name it `checkEquals`.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+      List<RowData> expected, List<RowData> actual, RowType rowType) {
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java]: Suggest to replace `RowType rowType` with `List<LogicalType> fieldTypes` to make it more general.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class RowDataTestUtils {
+
+  public static void checkEquals(RowData actual, RowData expected, RowType rowType) {
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java]: Suggest to replace `RowType rowType` with `List<LogicalType> fieldTypes` to make it more general.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    assertEquals("Row arity mismatch", expected.getArity(), actual.getArity());
+
+    List<LogicalType> fieldTypes = rowType.getChildren();
+    for (int i = 0; i < actual.getArity(); i++) {
+      RowData.FieldGetter getter = RowData.createFieldGetter(fieldTypes.get(i), i);
+      Object actualValue = getter.getFieldOrNull(actual);
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java:34]: Do we need to consider a field's logical type is not as expected for RowData? What will happen if getter is for string, but the data is int?

Comment (yuanhang-dev) [reply gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java:34]: Type mismatches will throw runtime exceptions. For example, in this scenario it throws:
`java.lang.ClassCastException: class org.apache.flink.table.data.binary.BinaryStringData cannot be cast to class java.lang.Integer`

This is actually the desired behavior for a test utility.

File: gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+      Object expectedValue = getter.getFieldOrNull(expected);
+
+      assertEquals("Field " + i + " mismatch", expectedValue, actualValue);
+    }
+  }
+
+  public static void checkEquals(
+      List<RowData> expected, List<RowData> actual, List<LogicalType> filedTypes) {
```
Comment (philo-he) [inline gluten-flink/ut/src/test/java/org/apache/gluten/streaming/api/operators/utils/RowDataTestUtils.java]: Use consistent argument order for actual and expected:
"(actual, expected, ..)"
